mailbox/memory/src/main/java/org/apache/james/mailbox/inmemory/InMemoryMessageIdManager.java

```diff
@@ -0,0 +1,164 @@
+                    .setFlags(newState, flagsUpdateMode, message.get().getUid().toRange(), mailboxSession);
+            }
+        }
+    }
+
+    @Override
+    public List<MessageResult> getMessages(List<MessageId> messages, FetchGroup fetchGroup, MailboxSession mailboxSession) throws MailboxException {
+        ImmutableList.Builder<MessageResult> builder = ImmutableList.builder();
+        List<MailboxMetaData> userMailboxes = mailboxManager.search(userMailboxes(mailboxSession), mailboxSession);
+        for (MailboxMetaData mailboxMetaData: userMailboxes) {
+            builder.addAll(retrieveMailboxMessages(mailboxMetaData.getId(), messages, fetchGroup, mailboxSession));
+        }
+        return builder.build();
+    }
+
+    private MailboxQuery userMailboxes(MailboxSession mailboxSession) {
+        return MailboxQuery.builder()
+                .matchesAll()
+                .username(mailboxSession.getUser().getUserName())
+                .build();
+    }
+
+    private List<MessageResult> retrieveMailboxMessages(MailboxId mailboxId, List<MessageId> messages, FetchGroup fetchGroup, MailboxSession mailboxSession) throws MailboxException {
+        ImmutableList.Builder<MessageResult> builder = ImmutableList.builder();
+        for (MessageId messageId: messages) {
+            Optional<MessageResult> maybeMessage = findMessageWithId(mailboxId, messageId, fetchGroup, mailboxSession);
+            if (maybeMessage.isPresent()) {
+                builder.add(maybeMessage.get());
+            }
+        }
+        return builder.build();
+    }
+
+    @Override
+    public void delete(MessageId messageId, List<MailboxId> mailboxIds, MailboxSession mailboxSession) throws MailboxException {
+        for (MailboxId mailboxId: mailboxIds) {
+            Optional<MessageResult> maybeMessage = findMessageWithId(mailboxId, messageId, FetchGroupImpl.MINIMAL, mailboxSession);
+            if (maybeMessage.isPresent()) {
+                MessageRange range = maybeMessage.get().getUid().toRange();
+                MessageManager messageManager = mailboxManager.getMailbox(mailboxId, mailboxSession);
+                messageManager.setFlags(new Flags(Flags.Flag.DELETED), FlagsUpdateMode.ADD, range, mailboxSession);
+                messageManager.expunge(range, mailboxSession);
+            }
+        }
+    }
+
+    @Override
+    public void setInMailboxes(MessageId messageId, List<MailboxId> mailboxIds, MailboxSession mailboxSession) throws MailboxException {
+        List<MessageResult> messages = getMessages(ImmutableList.of(messageId), FetchGroupImpl.MINIMAL, mailboxSession);
+        if (!messages.isEmpty()) {
+            ImmutableSet<MailboxId> currentMailboxes = currentMailboxes(messages).toSet();
+            
+            HashSet<MailboxId> targetMailboxes = Sets.newHashSet(mailboxIds);
+            List<MailboxId> mailboxesToRemove = ImmutableList.copyOf(Sets.difference(currentMailboxes, targetMailboxes));
```

Comment:
Ok, I did not understand it this way... My bad. Some changes are required on https://github.com/linagora/james-project/pull/507

(Unsure that we remove from mailbox)